### PR TITLE
Adds referral code read into browser

### LINKED ISSUE #12789: Read promo code from profile directory and set AppState
- There should be an argument passed during first run (on Windows there is; not sure about macOS after it's changed to PKG format). 
- During this first run, we should check for the presence of the `promoCode` file inside the user's profile directory (ex: `~/Library/Application Support/brave/promoCode`)
- Once file is detected and if code is valid, we need to set App State:
`state.setIn(['updates', 'promoCode'], promoCode)`

### LINKED ISSUE #12790: Contact promo API on first install and obtain download_id
**NOTE: https://github.com/brave/browser-laptop/issues/12789 must be completed first**
**NOTE: https://github.com/brave/browser-laptop/issues/12788 must be completed first**

During first run after install, the promo API needs to be contacted.

Example input/output from endpoint:
```
Input {
  "api_key": "key",
  "referral_code": "1234",
  "platform": "winx64"
}

Output {
  "download_id": "45f84b3d-6162-4565-ab44-dbed30f3ba6e",
  "referral_code": "1234"
}
```

The `download_id` returned should be stored in AppState

### PR DESCRIPTION
**Pull request by NejcZdovc**
Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).
- [x] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review). (Ask a Brave employee to help if you cannot access this document.)

Resolves #12789
Resolves #12790

Test Plan:
1) clean profile
2) create empty folder `brave-development` and create file `promoCode` with `123456` inside
3) start browser `LEDGER_ENVIRONMENT=staging npm run start`

Reviewer Checklist:

- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review) if one was not already requested.

Tests

- [ ] Adequate test coverage exists to prevent regressions
- [ ] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

### CODE REVIEW

File: app/browser/api/ledger.js [hunk 1]
```diff
@@ -81,6 +83,7 @@ let verifiedTimeoutId = false
 let v2RulesetDB
 const v2RulesetPath = 'ledger-rulesV2.leveldb'
 const statePath = 'ledger-state.json'
+const referralCodePath = 'promoCode/code'
```
Comment (petemill) [inline app/browser/api/ledger.js]: Please update to simply 'promoCode', or even better use the function for reading first-run promoCode from the module I included in #12920

Comment (NejcZdovc) [reply app/browser/api/ledger.js]: will change to your read when #12920 is merged

**Comment by diracdeltas** [comment]
please request a security review for this using the usual process when it's ready, thanks

File: app/browser/api/ledger.js (hunk 1 above)
Comment (NejcZdovc) [reply app/browser/api/ledger.js]: done

File: app/browser/api/ledger.js [hunk 2]
```diff
@@ -1852,7 +1868,63 @@ const onCallback = (state, result, delayTime) => {
...
+
+  if (code.length > 2) {
+    module.exports.roundtrip({
+      server: referralServer,
+      method: 'PUT',
+      path: '/promo/initialize/nonua',
+      payload: {
+        api_key: referralAPI,
+        referral_code: code,
+        platform: 'winx64' // TODO add correct platform
```
Comment (bbondy) [inline app/browser/api/ledger.js]: ^ seems blocking

File: app/browser/api/ledger.js [hunk 3]
```diff
@@ -1875,6 +1947,19 @@ const initialize = (state, paymentsEnabled) => {
     }
   }
 
+  if (!ledgerState.getInfoProp(state, 'referralTimestamp')) {
+    fs = require('fs')
```
[the remark this answers is not in the document]
Comment (NejcZdovc) [reply app/browser/api/ledger.js]: fs is defined locally, so not sure if there is any scenario where fs would exists?

File: app/browser/api/ledger.js (hunk 2 above)
Comment (NejcZdovc) [reply app/browser/api/ledger.js]: done

**Comment by NejcZdovc** [comment]
don't merge before #12920 is merged

File: app/browser/api/ledger.js [hunk 4]
```diff
@@ -95,6 +99,27 @@ const clientOptions = {
...
+  'linux': 'linux'
+}
+
+let platform = platforms[process.platform]
+
+if (process.platform === 'win32') {
+  platform = platforms[process.platform + process.arch]
+}
+
+let referralServer = 'https://laptop-updates-staging.herokuapp.com'
```
Comment (diracdeltas) [inline app/browser/api/ledger.js]: Privacy note: We are trying to not add any more 3rd party requests in Brave (https://github.com/brave/browser-laptop/issues/12632). For production, please make sure this endpoint is a brave.com subdomain, assigning a CNAME if needed.

cc @aekeus

File: app/browser/api/ledger.js [hunk 5]
```diff
@@ -95,6 +99,27 @@ const clientOptions = {
...
+let referralAPI = 'key'
+
+if (clientOptions.environment === 'production') {
+  referralServer = 'https://laptop-updates.brave.com'
```
Comment (diracdeltas) [inline app/browser/api/ledger.js]: ++, i think this addresses the comment above

File: app/browser/api/ledger.js [hunk 6]
```diff
@@ -1852,7 +1881,63 @@ const onCallback = (state, result, delayTime) => {
...
+    return
+  }
+
+  if (!code) {
+    return
+  }
+
+  code = code.trim()
+
+  if (code.length > 2) {
```
Comment (diracdeltas) [inline app/browser/api/ledger.js]: i thought the code was always going to be 6 characters? or can it be less?

File: app/browser/api/ledger.js [hunk 7]
```diff
@@ -2627,6 +2725,46 @@ const onPublisherTimestamp = (state, oldTimestamp, newTimestamp) => {
...
+  if (body && body.finalized) {
+    appActions.onReferralActivity()
+    return
+  }
+
+  if (clientOptions.verboseP) {
+    console.log('Referral is still not ready, please wait at least 30days')
+  }
+
+  updater.checkForUpdate(false, true)
```
Comment (diracdeltas) [inline app/browser/api/ledger.js]: minor: `activityRoundTrip` and `checkReferralActivity` would be clearer if they only handled things related to referral activity instead of also checking for browser updates. instead of returning after `appActions.checkReferralActivity()` in `updater.checkForUpdate`, you could always check for the browser update so that these functions here don't need to check for browser updates.

File: app/updater.js [hunk 8]
```diff
@@ -212,8 +214,21 @@ var downloadHandler = (err, metadata) => {
...
-  const updateStatus = AppStore.getState().getIn(['updates', 'status'])
+exports.checkForUpdate = (verbose, skipReferral = false) => {
+  const state = AppStore.getState()
+
+  // check for referral 30 days
+  if (!skipReferral && !ledgerState.getInfoProp(state, 'referralTimestamp')) {
+    const installTime = state.get('firstRunTimestamp')
+    const month = ledgerUtil.milliseconds.day * 30
+
+    if (installTime + month < new Date().getTime()) {
```
Comment (diracdeltas) [inline app/updater.js]: this function will check for referral activity every 30 days. i thought it was only supposed to check referral activity once after 30 days and then never check again.

File: docs/state.md [hunk 9]
```diff
@@ -217,8 +217,9 @@ AppStore
         EUR: number,
         USD: number
       },
-      reconcileFrequency: number // duration between each reconciliation in days
+      reconcileFrequency: number, // duration between each reconciliation in days
       reconcileStamp: number,  // timestamp for the next reconcilation
+      referralTimestamp: number,
```
Comment (diracdeltas) [inline docs/state.md]: please document this

File: docs/state.md [hunk 10]
```diff
@@ -600,6 +601,7 @@ AppStore
       name: string, // name of the update
       notes: string // release notes for the active update
     },
+    downloadId: string,
```
Comment (diracdeltas) [inline docs/state.md]: please document this

File: app/browser/api/ledger.js [hunk 11]
```diff
@@ -1875,6 +1960,19 @@ const initialize = (state, paymentsEnabled) => {
     }
   }
 
+  if (!ledgerState.getInfoProp(state, 'referralTimestamp')) {
+    fs = require('fs')
+    const codePath = pathName(referralCodePath)
+
+    fs.access(codePath, fs.FF_OK, (err) => {
+      if (err) {
```
Comment (diracdeltas) [inline app/browser/api/ledger.js]: this should set a flag in session state so that the referralCodePath check doesn't run every time a user starts Brave if we already know that they don't have a referral code. i guess you could just set `referralTimestamp` to a special value like 'none'.

File: app/browser/api/ledger.js [hunk 12]
```diff
@@ -1875,6 +1960,19 @@ const initialize = (state, paymentsEnabled) => {
     }
   }
 
+  if (!ledgerState.getInfoProp(state, 'referralTimestamp')) {
```
Comment (diracdeltas) [inline app/browser/api/ledger.js]: shouldn't this check if `downloadId` exists instead of `referralTimestamp`? IIUC, `referralTimestamp` is only set when the activity check is called, after 30 days, whereas `downloadId` is set as soon as the referral is initialized.

File: app/browser/api/ledger.js (hunk 6 above)
Comment (NejcZdovc) [reply app/browser/api/ledger.js]: done

File: app/browser/api/ledger.js (hunk 7 above)
Comment (NejcZdovc) [reply app/browser/api/ledger.js]: we want to reduce amount of calls, so we only want to call update server when referral fails.

File: app/updater.js (hunk 8 above)
Comment (NejcZdovc) [reply app/updater.js]: server can be in a different zone or not aligned with the browsers, so we need to check until referral is successful. We have `referralTimestamp` above which will not trigger this checks anymore as soon as we get green light from the server. We can add amount of attempts after which we will just stop checking?

File: docs/state.md (hunk 9 above)
Comment (NejcZdovc) [reply docs/state.md]: done

File: docs/state.md (hunk 10 above)
Comment (NejcZdovc) [reply docs/state.md]: done

File: app/browser/api/ledger.js (hunk 12 above)
Comment (NejcZdovc) [reply app/browser/api/ledger.js]: you are correct, fixed

File: app/browser/api/ledger.js (hunk 11 above)
Comment (NejcZdovc) [reply app/browser/api/ledger.js]: we have check above `if (!updateState.getUpdateProp(state, 'downloadId')) {`

Comment (diracdeltas) [reply app/browser/api/ledger.js]: @NejcZdovc but isn't `!updateState.getUpdateProp(state, 'downloadId')` always true for a user who didn't install from a referral? i mean that this code shouldn't keep running on every startup for those users.

File: app/updater.js (hunk 8 above)
Comment (diracdeltas) [reply app/updater.js]: Where is `referralTimestamp` set to a false value after getting a green light from the server?

Comment (NejcZdovc) [reply app/updater.js]: we don't set it to false, we actually set value (timestamp) to it in the ledger reducer

File: app/browser/api/ledger.js (hunk 11 above)
Comment (NejcZdovc) [reply app/browser/api/ledger.js]: that's correct, good catch

Comment (NejcZdovc) [reply app/browser/api/ledger.js]: But how would we know that we don't need to check for referral code?

Comment (NejcZdovc) [reply app/browser/api/ledger.js]: this file check is the only indicator that we have to know if install is from referral or not

Comment (NejcZdovc) [reply app/browser/api/ledger.js]: on thing that I can do is to set referral flag to false after we fail the first time. Because this should be only checked once that should be fine. wdyt?

Comment (NejcZdovc) [reply app/browser/api/ledger.js]: @diracdeltas

Comment (NejcZdovc) [reply app/browser/api/ledger.js]: ok did implement it b93829db3d5eaf7b0803625f225b87f5ac2b4748

Comment (NejcZdovc) [reply app/browser/api/ledger.js]: Implemented logic: we only check for that file once. And if we get error when preforming `fs.access` we will not check again

File: app/browser/api/ledger.js [hunk 13]
```diff
@@ -1852,7 +1881,63 @@ const onCallback = (state, result, delayTime) => {
...
+    return
+  }
+
+  if (!code) {
+    return
+  }
+
+  code = code.trim()
+
+  if (code.length > 0) {
```
Comment (diracdeltas) [inline app/browser/api/ledger.js]: not a big deal but why was this changed back to `> 0` from `=== 6`?

Comment (NejcZdovc) [reply app/browser/api/ledger.js]: because server is currently expecting code 1234 for testing, I change it to this

**Comment by petemill** [comment]
0.22.x f5eefc227
0.21.x b03ab808c
0.20.x 111b59b35
